lib/posix/nanosleep.c

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2018 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <limits.h>
+#include <errno.h>
+/* required for struct timespec */
+#include <posix/time.h>
+
+#define NS_PER_SEC 1000000000ULL
+#define NS_PER_USEC 1000ULL
+
+/**
+ * @brief Suspend execution for nanosecond intervals.
+ *
+ * See IEEE 1003.1
+ */
+int nanosleep(const struct timespec *rqtp, struct timespec *rmtp)
+{
+	u64_t ns;
+	const bool write = (rmtp != NULL);
+
+	if (rqtp == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+	if (rqtp->tv_sec < 0 || rqtp->tv_nsec < 0
+		|| rqtp->tv_nsec >= NS_PER_SEC) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (rqtp->tv_sec == 0 && rqtp->tv_nsec == 0) {
+		goto maybe_write;
+	}
+
+	if (unlikely(rqtp->tv_sec >= ULLONG_MAX / NS_PER_SEC)) {
+		/* If a user passes this in, we could be here a while, but
+		 * at least it's technically correct-ish
+		 */
+		ns = rqtp->tv_nsec + NS_PER_SEC;
+		k_sleep(K_SECONDS(rqtp->tv_sec - 1));
+	} else {
+		ns = rqtp->tv_sec * NS_PER_SEC + rqtp->tv_nsec;
+	}
+
+	/* currently we have no mechanism to achieve greater resolution */
+	k_busy_wait(ns / NS_PER_USEC);
```

Comment:
Hmm, could we still call `k_sleep()` as it supports us sleep values, not sure if it can actually sleep that short time.

Reply:
I assume the usecase for nanosleep() is really short intervals, which also should be as precise as possible, that's why @cfriedt used k_busy_wait(). k_sleep() leads to rescheduling of the thread, so any hopes for precision are lost.